### PR TITLE
Python unittest fix PR

### PR DESCRIPTION
**Pull request by swetha097**
Fixed Tf classification/ detection issue in GPU backend
Fixed RMN and Vignette black image issue

### CODE REVIEW

File: rocAL/rocAL_pybind/amd/rocal/plugin/tf.py [hunk 1]
```diff
@@ -54,11 +55,13 @@ def __iter__(self):
         return self
 
 class ROCALGenericIteratorDetection(object):
-    def __init__(self, pipeline, tensor_layout = types.NCHW, reverse_channels = False, multiplier = [1.0,1.0,1.0], offset = [0.0, 0.0, 0.0], tensor_dtype=types.FLOAT):
```
Comment (swetha097) [inline rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:57]: Please add a space after the ","

Comment (swetha097) [reply rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:57]: Please maintain space before and after "="

File: rocAL/rocAL_pybind/amd/rocal/plugin/tf.py [hunk 2]
```diff
@@ -54,11 +55,13 @@ def __iter__(self):
...
+    def __init__(self, pipeline, tensor_layout = types.NCHW, reverse_channels = False, multiplier = [1.0,1.0,1.0], offset = [0.0, 0.0, 0.0], tensor_dtype=types.FLOAT,device="cpu", device_id =0):
         self.loader = pipeline
         self.tensor_format =tensor_layout
         self.multiplier = multiplier
         self.offset = offset
+        self.device= device
```
Comment (swetha097) [inline rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:63]: Please maintain a space before and after "=" operator

File: rocAL/rocAL_pybind/amd/rocal/plugin/tf.py [hunk 3]
```diff
@@ -54,11 +55,13 @@ def __iter__(self):
...
+        self.device_id = device_id
```
Comment (swetha097) [inline rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:64]: Follow the above comment

File: rocAL/rocAL_pybind/amd/rocal/plugin/tf.py [hunk 4]
```diff
@@ -75,9 +78,17 @@ def __init__(self, pipeline, tensor_layout = types.NCHW, reverse_channels = Fals
             data_type="float16"
         
         if(types.NHWC == self.tensor_format):
-            self.out = np.zeros(( self.bs*self.n, int(self.h/self.bs), self.w, self.p), dtype = data_type)
-        else: 
-            self.out = np.zeros(( self.bs*self.n, self.p, int(self.h/self.bs), self.w), dtype = data_type)
+            if self.device == "cpu":
+                self.out = np.zeros(( self.bs*self.n, int(self.h/self.bs), self.w, self.p), dtype = data_type)
```
Comment (swetha097) [inline rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:82]: self.out = np.zeros(( self.bs*self.n, int(self.h/self.bs), self.w, self.p), dtype = data_type)
please format this line & below lines added, there seems to be unequal spacing

File: rocAL/rocAL_pybind/amd/rocal/plugin/tf.py [hunk 5]
```diff
@@ -192,10 +203,12 @@ def __init__(self,
...
                  fill_last_batch=True,
                  dynamic_shape=False,
-                 last_batch_padded=False):
+                 last_batch_padded=False,
+                 device="cpu",
+                 device_id =0):
         pipe = pipelines
         super(ROCALIterator, self).__init__(pipe, tensor_layout = pipe._tensor_layout, tensor_dtype = pipe._tensor_dtype,
-                                                            multiplier=pipe._multiplier, offset=pipe._offset)
+                                                            multiplier=pipe._multiplier, offset=pipe._offset,device=device, device_id=device_id)
```
Comment (swetha097) [inline rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:211]: Please maintain space after comma

File: rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py [hunk 6]
```diff
@@ -76,7 +79,7 @@ def main():
     # Build the pipeline
     pipe.build()
     # Dataloader
-    imageIterator = ROCALIterator(pipe)
+    imageIterator = ROCALIterator(pipe,device=device)
```
Comment (swetha097) [inline rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py:82]: Space after ,

File: rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py [hunk 7]
```diff
@@ -87,7 +90,7 @@ def main():
             print("\n\nPrinted first batch with", (batch_size), "images!")
         for element in list(range(batch_size)):
             cnt = cnt + 1
-            draw_patches(images_array[element],cnt)
+            draw_patches(images_array[element],cnt,device)
```
Comment (swetha097) [inline rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py:93]: Space after comma

File: rocAL/rocAL_pybind/examples/rocAL_api_tf_detection_pipeline.py [hunk 8]
```diff
@@ -46,10 +47,10 @@ def get_weights(num_bboxes):
 
     return weights_array
 
-def draw_patches(img, idx, bboxes):
-    #image is expected as a tensor, bboxes as numpy
+def draw_patches(img, idx, bboxes,device_type):
     import cv2
-    # image = img.detach().numpy()
+    if device_type == "gpu":
+        img= cp.asnumpy(img)
```
Comment (swetha097) [inline rocAL/rocAL_pybind/examples/rocAL_api_tf_detection_pipeline.py:53]: Give a space Before and after "="

File: rocAL/rocAL_pybind/examples/rocAL_api_tf_detection_pipeline.py [hunk 9]
```diff
@@ -114,7 +116,7 @@ def main():
         resized = fn.resize(decoded_images, resize_x=300, resize_y=300)
         pipe.set_outputs(resized)
     pipe.build()
-    imageIterator = ROCALIterator(pipe)
+    imageIterator = ROCALIterator(pipe,device=device)
```
Comment (swetha097) [inline rocAL/rocAL_pybind/examples/rocAL_api_tf_detection_pipeline.py:119]: maintain a space after ,

File: rocAL/rocAL_pybind/examples/rocAL_api_tf_detection_pipeline.py [hunk 10]
```diff
@@ -138,7 +140,7 @@ def main():
             processed_tensors = (features_dict, labels_dict)
             if args.print_tensor:
                 print("\nPROCESSED_TENSORS:\n", processed_tensors)
-            draw_patches(images_array[element],cnt,bboxes_array[element])
+            draw_patches(images_array[element],cnt,bboxes_array[element],device)
```
Comment (swetha097) [inline rocAL/rocAL_pybind/examples/rocAL_api_tf_detection_pipeline.py:143]: Space after comma

File: rocAL/rocAL_pybind/examples/READERS_TEST_FILE.sh [hunk 11]
```diff
@@ -168,7 +168,7 @@ if [[ rocAL_api_caffe_reader -eq 1 ]]; then
 
     # Mention dataset_path
     # Classification
-    data_dir=$ROCAL_DATA_PATH/caffe/classification/ilsvrc12_train_lmdb/
+    data_dir=$ROCAL_DATA_PATH/caffe/classification/
```
Comment (swetha097) [inline rocAL/rocAL_pybind/examples/READERS_TEST_FILE.sh:171]: Why was this changed ?

File: rocAL/rocAL_pybind/amd/rocal/fn.py [hunk 12]
```diff
@@ -297,10 +297,6 @@ def resize_mirror_normalize(*inputs, bytes_per_sample_hint=0, resize_min=0, resi
     kwargs_pybind = {"input_image0": inputs[0], "resize_min":resize_min, "resize_max":resize_max, "mean":mean, "std_dev":std,
                      "is_output": False, "mirror": mirror}
     rmn = b.ResizeMirrorNormalize(Pipeline._current_pipeline._handle ,*(kwargs_pybind.values()))
```
Comment (swetha097) [inline rocAL/rocAL_pybind/amd/rocal/fn.py:299]: Have we tested RMN after removing this ?

Comment (swetha097) [reply rocAL/rocAL_pybind/amd/rocal/fn.py:299]: Why are we removing this

File: rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py [hunk 13]
```diff
@@ -39,6 +41,7 @@ def main():
     # Args
     imagePath = args.image_dataset_path
     rocalCPU = False if args.rocal_gpu else True
+    device = "cpu" if rocalCPU else "gpu"
     batch_size = args.batch_size
```
Comment (sampath1117) [inline rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py:45]: Naming convention does not like it follows snake case style
rocALCPU and imagePath are in camel case but rest of them are in snake case
Please stick to one convention instead of 2

**Review by sampath1117 — COMMENTED** [review]
@Akilesh2 
There are many instances where space is not added before and after operator in this PR
Please go through all the files and make this change wherever applicable

File: rocAL/rocAL_pybind/amd/rocal/plugin/tf.py [hunk 14]
```diff
@@ -75,9 +78,17 @@ def __init__(self, pipeline, tensor_layout = types.NCHW, reverse_channels = Fals
...
+            if self.device == "cpu":
+                self.out = np.zeros(( self.bs*self.n, int(self.h/self.bs), self.w, self.p), dtype = data_type)
+            else:
+                with cp.cuda.Device(device=self.device_id):
+                    self.out = cp.zeros(( self.bs * self.n, int(self.h/self.bs), self.w, self.p), dtype = data_type)
+        else:
+            if self.device == "cpu":
+                self.out = np.zeros(( self.bs*self.n, self.p, int(self.h/self.bs), self.w), dtype = data_type)
+            else:
+                with cp.cuda.Device(device=self.device_id):
```
Comment (swetha097) [inline rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:90]: Have we tested tf gpu backend with these changes ?
How are we converting the cupy arrays to tf tensors , is this feasible ?

File: rocAL/rocAL_pybind/amd/rocal/plugin/tf.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:82]: self.bs*self.n - self.n denotes number of augmnetations @sampath1117 , this formula is correct

File: rocAL/rocAL_pybind/amd/rocal/fn.py [hunk 15]
```diff
@@ -297,10 +297,6 @@ def resize_mirror_normalize(*inputs, bytes_per_sample_hint=0, resize_min=0, resi
...
-    Pipeline._current_pipeline._tensor_layout = output_layout
-    Pipeline._current_pipeline._tensor_dtype = output_dtype
-    Pipeline._current_pipeline._multiplier = list(map(lambda x: 1/x ,std))
-    Pipeline._current_pipeline._offset = list(map(lambda x,y: -(x/y), mean, std))
     return (rmn)
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL_pybind/amd/rocal/fn.py:300]: This removed lines overwrites the tensor layout  and dtype passed to pipeline and overwrites  the multiplier and offset calculated the pipeline init()

File: rocAL/rocAL_pybind/examples/READERS_TEST_FILE.sh [hunk 16]
```diff
@@ -251,7 +251,7 @@ if [[ rocAL_api_caffe2_reader -eq 1 ]]; then
 
     # Mention dataset_path
     # Detection
-    data_dir=$ROCAL_DATA_PATH/caffe2/detection/lmdb_records/
+    data_dir=$ROCAL_DATA_PATH/caffe2/detection/
```
[the remark this answers is not in the document]
Comment (Akilesh2) [reply rocAL/rocAL_pybind/examples/READERS_TEST_FILE.sh:254]: In the latest MIvisionx-data, we don't have a subfolder inside the classification/detection. Inside classification/detection folder, we directly have mdb files, so I have made this path change

File: rocAL/rocAL_pybind/amd/rocal/plugin/tf.py [hunk 17]
```diff
@@ -19,6 +19,7 @@
 # THE SOFTWARE.
 
 import numpy as np
+import cupy as cp
 import rocal_pybind as b
```
[the remark this answers is not in the document]
Comment (Akilesh2) [reply rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:23]: I tested with 5.5 docker(swethabs/rocm5.5_pytorch_1.12.1_mivisionx_py3.8:v0.1) .It works fine

File: rocAL/rocAL_pybind/amd/rocal/plugin/tf.py (hunk 14 above)
Comment (Akilesh2) [reply rocAL/rocAL_pybind/amd/rocal/plugin/tf.py:90]: Yes, i have tested tf with GPU backend
I have made changes in the example file to handle cupy array

File: rocAL/rocAL_pybind/examples/READERS_TEST_FILE.sh [hunk 18]
```diff
@@ -196,7 +196,7 @@ if [[ rocAL_api_caffe_reader -eq 1 ]]; then
 
     # Mention dataset_path
     # Detection
-    data_dir=$ROCAL_DATA_PATH/caffe/detection/lmdb_record/
+    data_dir=$ROCAL_DATA_PATH/caffe/detection/
```
[the remark this answers is not in the document]
Comment (Akilesh2) [reply rocAL/rocAL_pybind/examples/READERS_TEST_FILE.sh:199]: In the latest MIvisionx-data, we don't have a subfolder inside the classification/detection. Inside classification/detection folder, we directly have mdb files, so I have made this path change

File: rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py [hunk 19]
```diff
@@ -26,70 +26,73 @@
...
     args = parse_args()
     # Args
-    imagePath = args.image_dataset_path
-    rocalCPU = False if args.rocal_gpu else True
+    image_path = args.image_dataset_path
+    rocal_cpu = False if args.rocal_gpu else True
+    device = "cpu" if rocal_cpu else "gpu"
     batch_size = args.batch_size
-    oneHotLabel = 1
+    one_hot_labels = 0
```
Comment (swetha097) [inline rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py:46]: Why have we changed one_hot_labels to 0 here ?

File: rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py [hunk 20]
```diff
@@ -26,70 +26,73 @@
...
     num_threads = args.num_threads
-    TFRecordReaderType = 0
-    featureKeyMap = {
+    TF_record_reader_type  = 0
```
Comment (swetha097) [inline rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py:48]: Make the first letter small case
